# Customer Behaviour and Produce Performance Analysis

## Data exploration

This notbook explores the structure, quality, and coverage of Olist e-commerce dataset.

The main objectives are:

- Understand the structure of each dataset
- Check data type
- Identify the missing values
- Identify the duplicate records
- Review order status
- Understand the available data range
- Prepare for data cleaning

In [1447]:
from pathlib import Path
import pandas as pd
import numpy as np


In [1448]:
data_path = Path("../data/raw/Dataset")
data_path.exists()

True

In [1449]:
customers = pd.read_csv(data_path / "olist_customers_dataset.csv")
orders = pd.read_csv(data_path / "olist_orders_dataset.csv")
order_items = pd.read_csv(data_path / "olist_order_items_dataset.csv")
products = pd.read_csv(data_path / "olist_products_dataset.csv")
reviews = pd.read_csv(data_path / "olist_order_reviews_dataset.csv")
category_translation = pd.read_csv(data_path / "product_category_name_translation.csv")

In [1450]:
# Set up a dictionary

datasets = {
    "customers": customers,
    "orders": orders,
    "order_items": order_items,
    "products": products,
    "reviews": reviews,
    "category_tanslation": category_translation
}

In [1451]:
for name, df in datasets.items():
    print(f"{name}: {df.shape[0]:,} rows, {df.shape[1]:,} columns")

customers: 99,441 rows, 5 columns
orders: 99,441 rows, 8 columns
order_items: 112,650 rows, 7 columns
products: 32,951 rows, 9 columns
reviews: 99,224 rows, 7 columns
category_tanslation: 71 rows, 2 columns


## Dataset size

This dataset contains separate tables for customers, orders, order items, products, and product reviews.

The orders and customers tables contain approx. the same number of rows because customer_id represents an order-level customer identifier. 

In [1452]:
# Check the first 5 lines of the dataset

customers.head()

# customer_id usually associated with single order.
# customer_unique_id can be used to identify whether a customer has made repeat purchases.

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [1453]:
orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [1454]:
order_items.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


In [1455]:
products.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In [1456]:
reviews.head()

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


In [1457]:
# Translate to English

category_translation.head()

,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


In [1458]:
for name, df in datasets.items():
    print(f"\n{name.title()}")
    print(df.columns.tolist())


Customers
['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']

Orders
['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']

Order_Items
['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value']

Products
['product_id', 'product_category_name', 'product_name_lenght', 'product_description_lenght', 'product_photos_qty', 'product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm']

Reviews
['review_id', 'order_id', 'review_score', 'review_comment_title', 'review_comment_message', 'review_creation_date', 'review_answer_timestamp']

Category_Tanslation
['product_category_name', 'product_category_name_english']


In [1459]:
# Check the type of data

orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [1460]:
for name, df in datasets.items():
    print(f"\n{name.title()}")
    print(df.dtypes)


Customers
customer_id                 object
customer_unique_id          object
customer_zip_code_prefix     int64
customer_city               object
customer_state              object
dtype: object

Orders
order_id                         object
customer_id                      object
order_status                     object
order_purchase_timestamp         object
order_approved_at                object
order_delivered_carrier_date     object
order_delivered_customer_date    object
order_estimated_delivery_date    object
dtype: object

Order_Items
order_id                object
order_item_id            int64
product_id              object
seller_id               object
shipping_limit_date     object
price                  float64
freight_value          float64
dtype: object

Products
product_id                     object
product_category_name          object
product_name_lenght           float64
product_description_lenght    float64
product_photos_qty            float64
product_weight

## Data types

This data columns in the orders table are currently stored as object values.

These columns will be converted to datetime during the data cleaning process.

Price and freight_value are stored as numerical values.

In [1461]:
for name, df in datasets.items():
    print(f"\n{name.title()}")

    missing_values = df.isnull().sum()
    missing_values = missing_values[missing_values > 0]

    if missing_values.empty:
        print("No missing values")
    else:
        print(missing_values.sort_values(ascending=False))


Customers
No missing values

Orders
order_delivered_customer_date    2965
order_delivered_carrier_date     1783
order_approved_at                 160
dtype: int64

Order_Items
No missing values

Products
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

Reviews
review_comment_title      87656
review_comment_message    58247
dtype: int64

Category_Tanslation
No missing values


In [1462]:
missing_summary = []

for name, df in datasets.items():
    for column in df.columns:
        missing_count = df[column].isnull().sum()

        if missing_count > 0:
            missing_summary.append({
                "datasets": name,
                "column": column,
                "missing_count": missing_count,
                "missing_percentage": round(missing_count / len(df) * 100, 2)
            })

missing_summary_df = pd.DataFrame(missing_summary)

missing_summary_df.sort_values(by="missing_percentage", ascending=False)

,datasets,column,missing_count,missing_percentage
11,reviews,review_comment_title,87656,88.34
12,reviews,review_comment_message,58247,58.70
2,orders,order_delivered_customer_date,2965,2.98
3,products,product_category_name,610,1.85
4,products,product_name_lenght,610,1.85
5,products,product_description_lenght,610,1.85
6,products,product_photos_qty,610,1.85
1,orders,order_delivered_carrier_date,1783,1.79
0,orders,order_approved_at,160,0.16
7,products,product_weight_g,2,0.01


In [1463]:
duplicate_summary = []

for name, df in datasets.items():
    duplicate_count = df.duplicated().sum()

    duplicate_summary.append({
        "dataset": name,
        "duplicate_row": duplicate_count
    })

duplicate_summary_df = pd.DataFrame(duplicate_summary)

duplicate_summary_df

,dataset,duplicate_row
0,customers,0
1,orders,0
2,order_items,0
3,products,0
4,reviews,0
5,category_tanslation,0


In [1464]:
# Check whether the keys are unique

key_checks = {
    "customers.customer_id": customers["customer_id"].is_unique,
    "orders.order_id": orders["order_id"].is_unique,
    "products.product_id": products["product_id"].is_unique,
    "category_translation.product_category_name": category_translation["product_category_name"].is_unique
}

pd.Series(key_checks, name = "is_unique")

customers.customer_id                         True
orders.order_id                               True
products.product_id                           True
category_translation.product_category_name    True
Name: is_unique, dtype: bool

In [1465]:
reviews["order_id"].value_counts().head(10)

order_id
c88b1d1b157a9999ce368f218a407141    3
8e17072ec97ce29f0e1f111e598b0c85    3
df56136b8031ecd28e200bb18e6ddb2e    3
03c939fd7fd3b38f8485a0f95798f1f6    3
5cb890a68b91b6158d69257e4e2bc359    2
2143393cca994a4b8235bc1d67ded772    2
25320e12b3d6e8f54f17389037588bba    2
0176a6846bcb3b0d3aa3116a9a768597    2
d9e44c3fd2ce16086619f299e92e12d8    2
21cb4961399b846b168d0358100f473b    2
Name: count, dtype: int64

In [1466]:
# Check the order status

order_status_summary = (
    orders["order_status"]
    .value_counts()
    .rename_axis("order_status")
    .reset_index(name = "number_of_orders")
)

order_status_summary

,order_status,number_of_orders
0,delivered,96478
1,shipped,1107
2,canceled,625
3,unavailable,609
4,invoiced,314
5,processing,301
6,created,5
7,approved,2


In [1467]:
# Check the order status percentage

order_status_summary["percentage"] = (
    order_status_summary["number_of_orders"] / len(orders) * 100
).round(2)

order_status_summary

,order_status,number_of_orders,percentage
0,delivered,96478,97.02
1,shipped,1107,1.11
2,canceled,625,0.63
3,unavailable,609,0.61
4,invoiced,314,0.32
5,processing,301,0.30
6,created,5,0.01
7,approved,2,0.00


In [1468]:
# Check the datetime

purchase_dates = pd.to_datetime(
    orders["order_purchase_timestamp"],
    errors = "coerce"
)

print("Earliest purchase date:", purchase_dates.min())
print("Latest purchase data:", purchase_dates.max())

Earliest purchase date: 2016-09-04 21:15:19
Latest purchase data: 2018-10-17 17:30:18


In [1469]:
# Check the purchase time in years

purchase_dates.dt.year.value_counts().sort_index()

order_purchase_timestamp
2016      329
2017    45101
2018    54011
Name: count, dtype: int64

In [1470]:
# Number of customers

print("Customers IDs:", customers["customer_id"].nunique())
print("Unique Customers:", customers["customer_unique_id"].nunique())

# Unique customers means the customers that have made repeat puchases.

Customers IDs: 99441
Unique Customers: 96096


In [1471]:
# Number of orders

print("Total orders:", orders["order_id"].nunique())

Total orders: 99441


In [1472]:
# Number of products

print("Total products:", products["product_id"].nunique())

Total products: 32951


In [1473]:
# Price of products

order_items[["price", "freight_value"]].describe()

,price,freight_value
count,112650.000000,112650.000000
mean,120.653739,19.990320
std,183.633928,15.806405
min,0.850000,0.000000
25%,39.900000,13.080000
50%,74.990000,16.260000
75%,134.900000,21.150000
max,6735.000000,409.680000


In [1474]:
# Distribution of reviews

reviews["review_score"].value_counts().sort_index()

review_score
1    11424
2     3151
3     8179
4    19142
5    57328
Name: count, dtype: int64

In [1475]:
# Average review score

reviews["review_score"].mean().round(2)

np.float64(4.09)

In [1476]:
# Check if Customers and Orders are connected

orders_customers_check = orders.merge(
    customers,
    on = "customer_id",
    how = "left",
    validate = "many_to_one"
)

orders_customers_check.shape

(99441, 12)

In [1477]:
orders_customers_check["customer_unique_id"].isnull().sum()

np.int64(0)

In [1478]:
# Order and Order_items

order_items["order_id"].isin(orders["order_id"]).mean()

np.float64(1.0)

In [1479]:
# Products and Order items

order_items["product_id"].isin(products["product_id"]).mean()

np.float64(1.0)

In [1480]:
# Reviews and Orders

reviews["order_id"].isin(orders["order_id"]).mean()

np.float64(1.0)

In [1481]:
# Dataset overview

overview = {
    "total_orders": orders["order_id"].nunique(),
    "unique_customers": customers["customer_id"].nunique(),
    "total_products": products["product_id"].nunique(),
    "total_order_items": len(order_items),
    "average_review_score": round(reviews["review_score"].mean(), 2),
    "earliest_purchases_date": purchase_dates.min(),
    "latest_purchases_date": purchase_dates.max()
}

pd.Series(overview, name = "value")

total_orders                             99441
unique_customers                         99441
total_products                           32951
total_order_items                       112650
average_review_score                      4.09
earliest_purchases_date    2016-09-04 21:15:19
latest_purchases_date      2018-10-17 17:30:18
Name: value, dtype: object

## Data Quality Summary

### Dataset coverage

- The dataset covers purchases from 2016-09-04 to 2018-10-17.
- It contains 99,441 orders.
- It contains 99,441 unique customers.
- It contains 32,951 orders.

### Missing values

- Some orders have missing delivery dates.
- Some products havemissing product category names.
- Many reviews do not have written comments.
- Missing review comments do not prevent review score analysis. 

### Duplicate records

- 0 fully duplicated rows were identified.
- Repeated order IDs in the order items table are expected because one order can contain multiple products.

### Data type issues

- Order date columns are currently stored as text.
- Date columns need to be converted to datetime.
- Product category names need to be translated to English.

### Decision for data cleaning

- Keep the orginal raw files unchanged.
- Convert date fields to datetime.
- Create delivery time and delivery delay fields.
- Translate product category names.
- Use customer_unique_id for repeat purchase analysis.
- Review non-delivered orders before product and delivery analysis.